docs/async/async-in-depth.md

```diff
@@ -0,0 +1,121 @@
+# Async In Depth
+
+By [Phillip Carter](https://github.com/cartermp)
+
+Writing I/O-Bound or CPU-Bound asynchronous code is simple using the `async` and `await` keywords, but what's happening under the covers is fairly complex.  The two key objects involved are `Task` and `Task<T>`.
+
+## Task and Task<T>
+
+Tasks are constructs used to represent operations running in the background.
+
+*   `Task` represents a single operation which does not return a value.
+*   `Task<T>` represents a single operation which returns a value of type `T`.
+
+Tasks are awaitable, meaning that using `await` will allow your application or service to perform useful work while the task is running by yielding control to its caller until the task is done. If you’re using `Task<T>`, the `await` keyword will additionally “unwrap” the value returned when the Task is complete.
+
+It’s important to reason about Tasks as abstractions of work happening in the background, and _not_ an abstraction over multithreading. In fact, unless explicitly started on a new thread via `Task.Run`, a Task will start on the current thread and delegate work to the Operating System.
+
+## Deeper Dive into Tasks for an I/O-Bound Operation
+
+Here’s a 10,000 foot view of what happens with a typical async I/O call:
+
+```csharp
+public async Task<string> GetHtml()
+{
+	var client = new HttpClient();
+	
+	// Execution is still synchronous here!
+	// The task handle "getHtmlTask" represents the active HTTP request.
+	var getHtmlTask = client.GetStringAsync("http://www.dotnetfoundation.org");
+	
+	// Execution of GetHtml() is yielded to the caller here!
+	var html = await getHtmlTask;
+	
+	return html;
+}
+```
+
+The call to `GetStringAsync` makes its way through the .NET libraries until it reaches a system interop call (such as `P/Invoke` on Windows). This eventually makes the proper System API call (such as `write` to a socket file descriptor on Linux). That System API call is then dealt with in the kernel, where the I/O request is sent to the proper subsystem. Although details about scheduling the work on the appropriate device driver are different for each OS, eventually an “incomplete task” signal will be sent from the device driver, bubbling its way back up to the .NET runtime. This will be converted into a `Task` or `Task<T>` by the runtime and returned to the calling method. When `await` is encountered, execuction is yielded and the system can go do something else.
+
+When the device driver has the data, it sends an interrupt which eventually allows the OS to bubble the result back up to the runtime, which will then queue up the result of the Task. Eventually execution will return to the method which called `GetStringAsync` at the `await`, and will “unwrap” the return value from the `Task<string>` which was being awaited. The method now has the result!
```

Comment:
I'll think of something more to say and, if needed, talk briefly about OS specifics.  I can't speak for Windows, but on Linux it does queue up the result into a queue that gets processed according to a configured policy, where it'll eventually bubble its way back up through system calls in kernel space and back up to the runtime.  Since Windows is a hybrid kernel it may handle things differently.